### PR TITLE
Remove xarray in favor of xt::xtensor<double, 2>

### LINKED ISSUE #471: Tabulate should check input shape from Python
To avoid memory errors

### PR DESCRIPTION
**Pull request by jorgensd**
This avoid hidden copying and reshaping when passing functions from xt::xarray -> xt::xtensor<double, 2> -> xt::xarray inside the tabulate code.

Only remaining use of xt::xarray is in basix::math::solve, where it probably also should be avoided. 
I will leave that to a separate PR as it is only called during construction of elements.

Resolves #471

### CODE REVIEW

File: cpp/basix/quadrature.cpp [hunk 1]
```diff
@@ -236,38 +236,37 @@ std::vector<double> compute_gauss_jacobi_points(double a, int m)
   return x;
 }
 //-----------------------------------------------------------------------------
-std::pair<xt::xarray<double>, std::vector<double>>
+std::pair<xt::xtensor<double, 2>, std::vector<double>>
 compute_gauss_jacobi_rule(double a, int m)
 {
   /// @note Computes on [-1, 1]
   std::vector<double> _pts = compute_gauss_jacobi_points(a, m);
```
Comment (mscroggs) [inline cpp/basix/quadrature.cpp:243]: We should rename this `pts`

File: cpp/basix/e-bubble.cpp [hunk 2]
```diff
@@ -80,22 +80,22 @@ FiniteElement basix::element::create_bubble(cell::type celltype, int degree,
...
-  const auto points
+  const xt::xtensor<double, 2> points
       = lattice::create(celltype, degree, lattice::type::equispaced, false);
   const std::size_t ndofs = points.shape(0);
   x[tdim].push_back(points);
 
   // Create coefficients for order (degree-1) vector polynomials
   xt::xtensor<double, 2> phi1;
-  xt::xtensor<double, 1> bubble;
+  xt::xtensor<double, 2> bubble({pts.shape(0), (std::size_t)1});
```
Comment (mscroggs) [inline cpp/basix/e-bubble.cpp:90]: I'd like to make this one 1D again

File: test/test_tensor_products.py [hunk 3]
```diff
@@ -52,7 +52,7 @@ def test_tensor_product_factorisation(cell_type, degree, element_type, element_a
 
             values2 = np.empty((element.dim, 1))
             for fs, perm in factors:
-                evals = [e.tabulate(d, [p])[d, 0, :, 0] for e, p, d in zip(fs, point, ds)]
+                evals = [e.tabulate(d, p.reshape(1, -1))[d, 0, :, 0] for e, p, d in zip(fs, point, ds)]
```
[the remark this answers is not in the document]
Comment (jorgensd) [reply test/test_tensor_products.py:55]: Closing as we discussed this on teams

File: cpp/basix/e-lagrange.cpp [hunk 4]
```diff
@@ -965,8 +965,8 @@ FiniteElement basix::element::create_lagrange(cell::type celltype, int degree,
           cell::num_sub_entities(celltype, i),
           xt::xtensor<double, 3>({0, 1, 0}));
     }
-
-    auto pt = lattice::create(celltype, 0, lattice_type, true, simplex_method);
+    const xt::xtensor<double, 2> pt
```
Comment (mscroggs) [inline cpp/basix/e-lagrange.cpp:968]: There's another `auto pt` on line 116 we should maybe also update?

File: cpp/basix/lattice.cpp [hunk 5]
```diff
@@ -18,58 +18,54 @@ using namespace basix;
...
   if (exterior)
   {
     throw std::runtime_error(
         "Chebyshev points including endpoints are not supported.");
   }
 
-  std::array<std::size_t, 1> s = {static_cast<std::size_t>(n - 1)};
-  xt::xtensor<double, 1> x(s);
+  std::array<std::size_t, 2> s
+      = {static_cast<std::size_t>(n - 1), (std::size_t)1};
```
Comment (mscroggs) [inline cpp/basix/lattice.cpp:59]: Would this still work if it was `      = {static_cast<std::size_t>(n - 1), 1};`?

File: cpp/basix/lattice.cpp [hunk 6]
```diff
@@ -78,60 +74,62 @@ xt::xtensor<double, 1> create_interval_gl(int n, bool exterior)
...
   if (n == 0)
-    return {0.5};
-  auto _pts = quadrature::get_gl_points(n - 1);
-
-  std::array<std::size_t, 1> s = {static_cast<std::size_t>(n - 1)};
-  xt::xtensor<double, 1> x(s);
+    return {{0.5}};
+  const xt::xtensor<double, 2> pts = quadrature::get_gl_points(n - 1);
+  std::array<std::size_t, 2> s
+      = {static_cast<std::size_t>(n - 1), (std::size_t)1};
```
Comment (mscroggs) [inline cpp/basix/lattice.cpp:80]: Also here

File: cpp/basix/lattice.cpp [hunk 7]
```diff
@@ -268,33 +265,35 @@ xt::xtensor<double, 2> create_tri_equispaced(int n, bool exterior)
   return p;
 }
 //-----------------------------------------------------------------------------
+// Warp points: see Hesthaven and Warburton, Nodal Discontinuous
```
[the remark this answers is not in the document]
Comment (jorgensd) [reply cpp/basix/lattice.cpp:268]: No, that would indicate that `doxygen` should render them. As they are in an anonymous namespace that is not right

File: cpp/basix/quadrature.cpp [hunk 8]
```diff
@@ -256,18 +254,19 @@ compute_gauss_jacobi_rule(double a, int m)
     wts[i] = a1 / (1.0 - x * x) / (f * f);
   }
 
-  return {pts, wts};
+  return {xt::adapt(pts, {pts.size(), (std::size_t)1}), wts};
```
[the remark this answers is not in the document]
Comment (jorgensd) [reply cpp/basix/quadrature.cpp:257]: Inside adapt and reshape view, explicit casting to std::size_t is required.

**Pull request merged**